### PR TITLE
LRA messaging hooks for selected signatures

### PR DESCRIPTION
**Pull request by danielkec**
Signed-off-by: Daniel Kec <daniel.kec@oracle.com>

### CODE REVIEW

File: microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/HookTest.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            MessageContext.lookup(message).put("txId", idSeq.incrementAndGet());
+            Long txId = (Long) MessageContext.lookup(message).get("txId");
+            System.out.println("TxId:" + txId + " Invoking method " + method.getName());
+        }, (method, o) -> {
+            System.out.println("output "+o);
+            Long txId = (Long) MessageContext.lookup((Message<?>) o).get("txId");
+            System.out.println("TxId:" + txId + " Invoked method " + method.getName());
+        });
+    }
+
```
Comment (danielkec) [inline microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/HookTest.java:68]: @paulparkinson Idea is that I am trying to keep context hashmap between channels with Messages itself as keys(so context can be discarded by GC when message is consumed and not referenced any more). It is possible only for some signatures like in the example:
```java 
    @Outgoing("channel1")
    public PublisherBuilder<Message<String>> publish() {
        return ReactiveStreams.of("test1", "test2", "test3")
                .map(Message::of);
    }

    @Incoming("channel1")
    @Outgoing("channel2")
    public String process(String incoming) {
        return incoming.toUpperCase();
    }

    @Incoming("channel2")
    public void consume(String incoming) {
        System.out.println(">>>" + incoming);
    }
```
Output of this POC test look like this:
```
registering
TxId:1 Invoking method process
output org.eclipse.microprofile.reactive.messaging.Message$1@4ebea12c
TxId:1 Invoked method process
TxId:2 Invoking method consume
>>>TEST1
output org.eclipse.microprofile.reactive.messaging.Message$1@44c79f32
TxId:2 Invoked method consume
TxId:3 Invoking method process
output org.eclipse.microprofile.reactive.messaging.Message$1@757d6814
TxId:3 Invoked method process
TxId:4 Invoking method consume
>>>TEST2
output org.eclipse.microprofile.reactive.messaging.Message$1@649725e3
TxId:4 Invoked method consume
TxId:5 Invoking method process
output org.eclipse.microprofile.reactive.messaging.Message$1@4c168660
TxId:5 Invoked method process
TxId:6 Invoking method consume
>>>TEST3
output org.eclipse.microprofile.reactive.messaging.Message$1@52b56a3e
TxId:6 Invoked method consume
```

Comment (danielkec) [reply microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/HookTest.java:68]: hm its not right

File: microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/HookTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+public class HookTest {
+
+    @Inject
+    MessagingCdiExtension messagingCdiExtension;
+
+    AtomicLong idSeq = new AtomicLong();
+
+    private void makeConnections(@Observes @Priority(PLATFORM_AFTER + 80) @Initialized(ApplicationScoped.class) Object event,
+                                 BeanManager beanManager) {
+        System.out.println("registering");
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/HookTest.java:56]: Please remove all instances of `System.out.println` from your code. Even in tests.

**Review by tomas-langer — CHANGES_REQUESTED** [review]

**Review by paulparkinson — APPROVED** [review]
yes, this should do the trick.

**Review by paulparkinson — APPROVED** [review]
I have confirmed this is working well and can be merged when ready.  Thanks!

**Review by paulparkinson — APPROVED** [review]

**Review by paulparkinson — DISMISSED** [review]
this works well, thanks

File: messaging/connectors/jms/src/main/java/io/helidon/messaging/connectors/jms/JmsConnector.java [hunk 3]
```diff
@@ -409,7 +409,7 @@ private void produce(
 
                         if (m instanceof OutgoingJmsMessage) {
                             // custom mapping, properties etc.
-                            jmsMessage = ((OutgoingJmsMessage<?>) m).toJmsMessage(session, mapper.get());
+                            jmsMessage = ((OutgoingJmsMessageImpl<?>) m).toJmsMessage(session, mapper.get());
```
Comment (tomas-langer) [inline messaging/connectors/jms/src/main/java/io/helidon/messaging/connectors/jms/JmsConnector.java:412]: Seems like a brave cast - are we 100% sure this is our implementation?

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/AbstractMessagingMethod.java [hunk 4]
```diff
@@ -72,18 +79,23 @@ void validate() {
     }
 
     void init(BeanManager beanManager, Config config) {
+        this.config = config;
         this.beanInstance = ChannelRouter.lookup(bean, beanManager);
     }
 
-    Method getMethod() {
+    public String getName() {
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/AbstractMessagingMethod.java:86]: Why are these methods changed to `public`? Also if this is expected to be exposed by a `public` subclass, please add javadocs.

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/MessageContext.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+/**
+ * JVM wide context for every referenced message.
+ */
+public class MessageContext extends ConcurrentHashMap<String, Object> {
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/MessageContext.java:29]: Extending a map is not recommended in general. This is a strange construct - is there a design document that would explain this class and its usage?

**Review by tomas-langer — CHANGES_REQUESTED** [review]

**Comment by romain-grecourt** [comment]
closing for inactivity

**Pull request closed without merging**